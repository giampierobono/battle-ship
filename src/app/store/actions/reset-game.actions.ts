import { createAction } from '@ngrx/store';

export const resetGame = createAction('[BS webapp] Reset game');
