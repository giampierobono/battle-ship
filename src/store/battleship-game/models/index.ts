export * from './set-game-player-turn-payload.model';
export * from './save-player-move-payload.model';
export * from './set-winner-player-payload.model';
