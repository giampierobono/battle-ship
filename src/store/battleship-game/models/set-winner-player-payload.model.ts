export interface SetWinnerPlayerPayloadModel {
  winnerPlayerIndex: number;
}
