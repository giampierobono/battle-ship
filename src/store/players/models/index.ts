export * from './add-players-payload.model';
export * from './create-player-payload.model';
