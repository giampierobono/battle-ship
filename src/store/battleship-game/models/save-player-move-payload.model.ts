import { PositionStateEnum, ShotPositionModel } from '../../../models';

export interface SavePlayerMovePayloadModel {
  shotPosition: ShotPositionModel;
  state: PositionStateEnum;
}
