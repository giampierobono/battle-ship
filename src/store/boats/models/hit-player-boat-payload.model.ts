import { ShotPositionModel } from '../../../models';

export interface HitPlayerBoatPayloadModel {
  boatIndex: number;
  position: ShotPositionModel;
}
