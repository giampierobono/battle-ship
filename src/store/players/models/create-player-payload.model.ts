import { PlayerModel } from '../../../models';

export interface CreatePlayerPayloadModel {
  newPlayer: PlayerModel;
}
