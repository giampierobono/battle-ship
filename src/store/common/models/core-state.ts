import { IPlayersState } from '../../players';
import { IBoatsState } from '../../boats';
import { IBoatShapeState } from '../../boats-shapes';
import { IGameSettingsState } from '../../game-settings';
import { IBattleShipGameState } from '../../battleship-game';

export interface CoreState {
  players: IPlayersState;
  boats: IBoatsState;
  boatShapes: IBoatShapeState;
  settings: IGameSettingsState;
  game: IBattleShipGameState;
}
