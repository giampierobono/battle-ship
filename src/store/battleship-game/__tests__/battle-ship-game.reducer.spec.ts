import { battleShipGameInitialState, battleShipGameReducer } from '../battle-ship-game.reducer';
import { addSunkBoat, savePlayerMove, setPlayerTurn, setWinnerPlayer } from '../battle-ship-game.actions';
import { PositionStateEnum } from '../../../models';

describe('battleShipGameReducer', () => {
  it('should return current status on unknown action', () => {
    expect(battleShipGameReducer(battleShipGameInitialState, { type: 'UNKNOWN' })).toEqual(battleShipGameInitialState);
  });

  it('should set player turn to initial state', () => {
    const newState = battleShipGameReducer(battleShipGameInitialState, setPlayerTurn({ playerIndex: 1 }));
    expect(newState.playerTurn).toBe(1);
  });

  it('should save player move and set turn over', () => {
    const newState = battleShipGameReducer(
      battleShipGameInitialState,
      savePlayerMove({ state: PositionStateEnum.HIT, shotPosition: { playerIndex: 0, row: 0, column: 0 } })
    );
    expect(newState.playersMoves[0][0][0]).toBe(PositionStateEnum.HIT);
    expect(newState.isCurrentTurnOver).toBeTruthy();
  });

  it('should add sunk boat to player 1', () => {
    const newState = battleShipGameReducer(
      battleShipGameInitialState,
      addSunkBoat({ boatIndex: 0, boatBelongsToPlayerIndex: 0, sunkByPlayerIndex: 1 })
    );
    expect(newState.playersSunkBoats[1].length).toBe(1);
    expect(newState.playersSunkBoats[1]).toMatchSnapshot();
  });

  it('should set winner player', () => {
    const newState = battleShipGameReducer(battleShipGameInitialState, setWinnerPlayer({ winnerPlayerIndex: 1 }));
    expect(newState.winnerPlayer).toBe(1);
  });
});
