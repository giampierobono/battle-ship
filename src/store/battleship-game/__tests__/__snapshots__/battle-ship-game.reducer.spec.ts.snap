// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`battleShipGameReducer should add sunk boat to player 1 1`] = `
Array [
  Object {
    "boatBelongsToPlayerIndex": 0,
    "boatIndex": 0,
    "sunkByPlayerIndex": 1,
  },
]
`;
